src/libostree/ostree-repo.c

```diff
@@ -2659,6 +2659,48 @@ get_remotes_d_dir (OstreeRepo          *self,
     return g_file_resolve_relative_path (sysroot, SYSCONF_REMOTES);
 }
 
+static gboolean
+min_free_space_size_validate_and_convert (OstreeRepo    *self,
+                                          const char    *min_free_space_size_str,
+                                          GError       **error)
+{
+  static GRegex *regex;
+  static gsize regex_initialized;
+  if (g_once_init_enter (&regex_initialized))
+    {
+      regex = g_regex_new ("^([0-9]+)(G|M|T)B$", 0, 0, NULL);
+      g_assert (regex);
+      g_once_init_leave (&regex_initialized, 1);
+    }
+
+  g_autoptr(GMatchInfo) match = NULL;
+  if (!g_regex_match (regex, min_free_space_size_str, 0, &match))
+    return glnx_prefix_error (error, "Failed to parse min-free-space-size parameter: '%s'", min_free_space_size_str);
+
+  g_autofree char *size_str = g_match_info_fetch (match, 1);
+  g_autofree char *unit = g_match_info_fetch (match, 2);
+  guint shifts;
+
+  switch (*unit)
+    {
+      case 'M':
+        shifts = 0;
+        break;
+      case 'G':
+        shifts = 10;
+        break;
+      case 'T':
+        shifts = 20;
+        break;
+      default:
+        g_assert_not_reached ();
+    }
+
+  self->min_free_space_mb = g_ascii_strtoull (size_str, NULL, 10) << shifts;
```

Comment:
Right. I will fix the documentation.
OTOH, can you tell me what should be behavior in case there is an overflow or a `g_assert` will do the trick?

Reply:
An overflow could be triggered by the user (because they write the config file), so has to be handled using a `GError`. `g_assert()` should only be used to check for programmer errors which can’t possibly be triggered by user input.

Reply:
hmm, Do you mean whether to check for wrapping around? As `self->min_free_space_mb` is already `guint64` hence it won't "overflow" (rather wrap around). Not sure how to detect that? :thinking:

Reply:
It will overflow, in the sense that if you want to fully represent a 64-bit integer shifted left by 20 bits, you need 84 bits of storage. You can only detect it before doing the operation:
```
guint64 min_free_space = g_ascii_strtoull (size_str, NULL, 10);
if (shifts > 0 && g_bit_nth_lsf (min_free_space, 64 - shifts) != -1)
  {
    g_set_error (…);
    return FALSE;
  }
self->min_free_space_mb = min_free_space << shifts;
```

That test goes by the principle that the shift will overflow if any of the `shifts` MSBs in `min_free_space` are set. I haven’t tested that, but it should be roughly correct.